scrapy_po/middleware.py

```diff
@@ -33,32 +37,39 @@ def process_response(self, request: Request, response, spider):
 
         raise returnValue(response)
 
-    @classmethod
-    def is_page_object(cls, argument_type) -> bool:
-        return issubclass(argument_type, PageObject)
-
 
 @inlineCallbacks
-def create_page_object(page_cls, response):
+def create_page_object(page_cls, response, dependency_graph=None):
     """
     Create PageObject instances, resolving all constructor dependencies.
     """
     # find out what resources are available
     provider_funcs = {cls: provider(response)
                       for cls, provider in providers.items()}
 
+    def can_provide(cls):
+        return cls in provider_funcs or is_page_object(cls)
+
     # find out which arguments page object needs.
     arguments = andi.inspect(page_cls.__init__)
-    kwargs_to_provide = andi.to_provide(arguments, can_provide=provider_funcs)
+    kwargs_to_provide = andi.to_provide(arguments, can_provide=can_provide)
     not_supported = arguments.keys() - kwargs_to_provide.keys()
     if not_supported:
         raise TypeError("Can't instantiate arguments: %s" % not_supported)
 
     # instantiate all arguments
     kwargs = {}
+    dependency_graph = dependency_graph or []
```

Comment:
I wonder if obtaining of a full dependency graph can be moved to andi, or separated in some way. We may need this outside Scrapy as well.

Reply:
Good idea. I have implemented a tentative version for andi here: https://github.com/scrapinghub/andi/pull/7 and a new implementation to replace current PR here: https://github.com/scrapinghub/scrapy-po/pull/5 . I will close this PR.